### PR TITLE
PDO tests for retrieving different data types as output params using various PDO param types

### PR DESCRIPTION
**Pull request by yitam**
This change is [Reviewable](https://reviewable.io/reviews/microsoft/msphpsql/704)

### CODE REVIEW

File: test/functional/pdo_sqlsrv/pdo_ae_output_param_binary_size.phpt [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+                $ord1 = convert2Hex($ch1, $length);
+                insertRow($conn, $tbname, array("c_det" => new BindParamOp(1, $input0, "PDO::PARAM_LOB", 0, "PDO::SQLSRV_ENCODING_BINARY"),
+                                                "c_rand" => new BindParamOp(2, $input1, "PDO::PARAM_LOB", 0, "PDO::SQLSRV_ENCODING_BINARY")), "prepareBindParam");
+                    
+                // fetch with PDO::bindParam using a stored procedure
+                dropProc($conn, $spname);
+                $spSql = "CREATE PROCEDURE $spname (
+                                @c_det $type OUTPUT, @c_rand $type OUTPUT ) AS
+                                SELECT @c_det = c_det, @c_rand = c_rand FROM $tbname";
+                $conn->query($spSql);
```
Comment (v-kaywon) [inline test/functional/pdo_sqlsrv/pdo_ae_output_param_binary_size.phpt]: there's a CreateProc function you can use

File: test/functional/pdo_sqlsrv/pdo_ae_output_param_binary_size.phpt [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+                
+                // call stored  procedure
+                $outSql = getCallProcSqlPlaceholders($spname, 2);
+                foreach ($pdoParamTypes as $pdoParamType) {
+                    $stmt = $conn->prepare($outSql);
+                    trace("\nParam $pdoParamType with INOUT = $inout\n");
+                    
+                    if ($inout && $pdoParamType != PDO::PARAM_STR) {
+                        // Currently do not support getting binary as strings + INOUT param
```
Comment (v-kaywon) [inline test/functional/pdo_sqlsrv/pdo_ae_output_param_binary_size.phpt]: we don't?! why?

File: test/functional/pdo_sqlsrv/pdo_ae_output_param_binary_size.phpt [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+                    $det = "";
+                    $rand = "";
+                    if ($pdoParamType == PDO::PARAM_STR) {
+                        $stmt->bindParam(1, $det, $paramType, $length, PDO::SQLSRV_ENCODING_BINARY); 
+                        $stmt->bindParam(2, $rand, $paramType, $length, PDO::SQLSRV_ENCODING_BINARY); 
+                    } elseif ($pdoParamType == PDO::PARAM_BOOL || $pdoParamType == PDO::PARAM_INT) {
+                        $det = $rand = 0;
+                        $stmt->bindParam(1, $det, $paramType, PDO::SQLSRV_PARAM_OUT_DEFAULT_SIZE);
+                        $stmt->bindParam(2, $rand, $paramType, PDO::SQLSRV_PARAM_OUT_DEFAULT_SIZE);
+                    } else {
```
Comment (v-kaywon) [inline test/functional/pdo_sqlsrv/pdo_ae_output_param_binary_size.phpt]: I thought PDO::SQLSRV_ENCODING_BINARY can be specified with PDO::PARAM_LOB

File: test/functional/pdo_sqlsrv/pdo_ae_output_param_binary_size.phpt [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+                        } elseif ($pdoParamType == PDO::PARAM_BOOL) {
+                            // for boolean values, they should all be bool(true)
+                            // because all floats are non-zeroes
+                            // This only occurs without AE
+                            // With AE enabled, this would have caused an exception
+                            if (!$det || !$rand) {
+                                printValues($errMsg, $det, $rand, $input0, $input1);
+                            }
+                        } else {
+                            // $pdoParamType is PDO::PARAM_INT
```
Comment (v-kaywon) [inline test/functional/pdo_sqlsrv/pdo_ae_output_param_binary_size.phpt]: what about PDO::PARAM_LOB and PDO::PARAM_NULL?

File: test/functional/pdo_sqlsrv/pdo_ae_output_param_char_size.phpt [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+                                    printValues($errMsg, $det, $rand, $input0, $input1);
+                                }
+                            }
+                        } elseif ($det !== $input0 || $rand !== $input1) {
+                            printValues($errMsg, $det, $rand, $input0, $input1);
+                        }
+                    } catch (PDOException $e) {
+                        $message = $e->getMessage();
+                        $errMsg = "EXCEPTION: ****$type as $pdoParamType failed with INOUT = $inout:****\n";
+                        if ($pdoParamType == PDO::PARAM_NULL || $pdoParamType == PDO::PARAM_LOB) {
```
Comment (v-kaywon) [inline test/functional/pdo_sqlsrv/pdo_ae_output_param_char_size.phpt]: I'm surprise PARAM_LOB doesn't work. I always thought the driver treat PARAM_STR and PARAM_LOB the same

File: test/functional/pdo_sqlsrv/pdo_ae_output_param_char_size.phpt [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+                        $paramType = $pdoParamType | PDO::PARAM_INPUT_OUTPUT;
+                    } else {
+                        $paramType = $pdoParamType;
+                    }
+                    
+                    $len = $length;
+                    if ($pdoParamType == PDO::PARAM_BOOL || $pdoParamType == PDO::PARAM_INT) {
+                        $len = PDO::SQLSRV_PARAM_OUT_DEFAULT_SIZE;
+                        $det = $rand = 0;
+                    }
```
Comment (v-kaywon) [inline test/functional/pdo_sqlsrv/pdo_ae_output_param_char_size.phpt]: so based on the PDO::PARAM_ type, you initialize the output variable differently. Then maybe you should also initialize $det = $rand = NULL for PDO::PARAM_NULL

File: test/functional/pdo_sqlsrv/pdo_ae_output_param_datetimes.phpt [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+                }
+
+                $type = "$dataType($precision)";
+                trace("\nTesting $type:\n");
+
+                //create and populate table
+                $colMetaArr = array(new ColumnMeta($type, "c_det"), new ColumnMeta($type, "c_rand", null, "randomized"));
+                createTable($conn, $tbname, $colMetaArr);
+
+                $stmt = insertRow($conn, $tbname, array("c_det" => $inputValues[0], "c_rand" => $inputValues[1] ), null, $r);
```
Comment (v-kaywon) [inline test/functional/pdo_sqlsrv/pdo_ae_output_param_datetimes.phpt]: don't think you need $r here

File: test/functional/pdo_sqlsrv/pdo_ae_output_param_datetimes.phpt [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+                    try {
+                        $stmt->execute();
+                        $errMsg = "****$type as $pdoParamType failed with INOUT = $inout:****\n";
+                        // What follows only happens with OUTPUT parameter
+                        if ($inout) {
+                            echo "Any datetime type as INOUT param should have caused an exception!\n";
+                        }
+                        if ($pdoParamType == PDO::PARAM_INT) {
+                            // Expect an integer, the first part of the date time string
+                            $ch = ($dataType == "time")? ':' : '-';
```
Comment (v-kaywon) [inline test/functional/pdo_sqlsrv/pdo_ae_output_param_datetimes.phpt]: what happens to datetime2 and datetimeoffset

File: test/functional/pdo_sqlsrv/pdo_ae_output_param_binary_size.phpt (hunk 2 above)
Comment (yitam) [reply test/functional/pdo_sqlsrv/pdo_ae_output_param_binary_size.phpt]: yes, created VSO 2829 for this

File: test/functional/pdo_sqlsrv/pdo_ae_output_param_binary_size.phpt (hunk 3 above)
Comment (yitam) [reply test/functional/pdo_sqlsrv/pdo_ae_output_param_binary_size.phpt]: Not sure what you mean?

File: test/functional/pdo_sqlsrv/pdo_ae_output_param_binary_size.phpt (hunk 4 above)
Comment (yitam) [reply test/functional/pdo_sqlsrv/pdo_ae_output_param_binary_size.phpt]: they would have thrown exceptions

File: test/functional/pdo_sqlsrv/pdo_ae_output_param_char_size.phpt (hunk 6 above)
Comment (yitam) [reply test/functional/pdo_sqlsrv/pdo_ae_output_param_char_size.phpt]: But binding output as PARAM NULL always throws exception

File: test/functional/pdo_sqlsrv/pdo_ae_output_param_char_size.phpt (hunk 5 above)
Comment (yitam) [reply test/functional/pdo_sqlsrv/pdo_ae_output_param_char_size.phpt]: No they are not treated the same when binding as output parameters

File: test/functional/pdo_sqlsrv/pdo_ae_output_param_datetimes.phpt (hunk 8 above)
Comment (yitam) [reply test/functional/pdo_sqlsrv/pdo_ae_output_param_datetimes.phpt]: are you asking about inout parameter? Binding any date / time types as inout throw exceptions

File: test/functional/pdo_sqlsrv/pdo_ae_output_param_decimals.phpt [hunk 9]
```diff
@@ -0,0 +1,241 @@
...
+// are different from the expected inputs 
+function compareIntegers($det, $rand, $inputValues, $pdoParamType)
+{
+    ///////////////////////////////////////////////////////////////////////
+    // See GitHub issue 707 - Fix this method when the problem is addressed
+    //
+    // Assume $pdoParamType is PDO::PARAM_BOOL or PDO::PARAM_INT
+    if (is_string($det)) {
+        return (!compareFloats(floatval($det), $inputValues[0]) 
+                && !compareFloats(floatval($rand), $inputValues[1]));
```
Comment (v-kaywon) [inline test/functional/pdo_sqlsrv/pdo_ae_output_param_decimals.phpt]: when would $det be a string when you specify as PARAM_BOOL or PARAM_INT? so it returns a string with decimal places?

File: test/functional/pdo_sqlsrv/pdo_ae_output_param_decimals.phpt [hunk 10]
```diff
@@ -0,0 +1,241 @@
...
+                    }
+                    
+                    $type = "$dataType($precision, $scale)";
+                    trace("\nTesting $type:\n");
+                    
+                    //create and populate table
+                    $colMetaArr = array(new ColumnMeta($type, "c_det"), new ColumnMeta($type, "c_rand", null, "randomized"));
+                    createTable($conn, $tbname, $colMetaArr);
+
+                    $stmt = insertRow($conn, $tbname, array("c_det" => $inputValues[0], "c_rand" => $inputValues[1] ), null, $r);
```
Comment (v-kaywon) [inline test/functional/pdo_sqlsrv/pdo_ae_output_param_decimals.phpt]: $r is not needed

File: test/functional/pdo_sqlsrv/pdo_ae_output_param_decimals.phpt [hunk 11]
```diff
@@ -0,0 +1,241 @@
...
+                    // call stored procedure
+                    $outSql = getCallProcSqlPlaceholders($spname, 2);
+                    foreach ($pdoParamTypes as $pdoParamType) {
+                        $det = $rand = 0.0;
+                        $stmt = $conn->prepare($outSql);
+                    
+                        $len = 2048;
+                        if ($pdoParamType == PDO::PARAM_BOOL || $pdoParamType == PDO::PARAM_INT) {
+                            $len = PDO::SQLSRV_PARAM_OUT_DEFAULT_SIZE;
+                            $det = $rand = 0;
```
Comment (v-kaywon) [inline test/functional/pdo_sqlsrv/pdo_ae_output_param_decimals.phpt]: perhaps for PDO::PARAM_STR and PDO::PARAM_LOB you want to specify $det = $rand = ""; as well

File: test/functional/pdo_sqlsrv/pdo_ae_output_param_decimals.phpt [hunk 12]
```diff
@@ -0,0 +1,241 @@
...
+// this function returns true if the floats are more different than expected
+function compareFloats($actual, $expected) 
+{
+    $epsilon = 0.00001;
+    $diff = abs(($actual - $expected) / $expected);
+    return ($diff > $epsilon);
+}
+
+// function compareDecimals() returns false when the fetched values 
+// are different from the expected inputs
```
Comment (v-kaywon) [inline test/functional/pdo_sqlsrv/pdo_ae_output_param_decimals.phpt]: change the comment to compareIntegers() blah blah

File: test/functional/pdo_sqlsrv/pdo_ae_output_param_nchar_size.phpt [hunk 13]
```diff
@@ -0,0 +1,173 @@
...
+                        $det = $rand = 0;
+                    }
+                    
+                    $stmt->bindParam(1, $det, $paramType, $len); 
+                    $stmt->bindParam(2, $rand, $paramType, $len); 
+                    
+                    try {
+                        $stmt->execute();
+                        $errMsg = "****$type as $pdoParamType failed with INOUT = $inout:****\n";
+                        if ($length < 64 && $pdoParamType != PDO::PARAM_STR) {
```
Comment (v-kaywon) [inline test/functional/pdo_sqlsrv/pdo_ae_output_param_nchar_size.phpt]: maybe put a comment saying when the length is greater than 64, a string is returned regardless of the PDO PARAM type.

**Comment by v-kaywon** [comment]
done reviewing, but I still don't understand why PARAM_LOB doesn't work for all the types. Do you have any documentation on that?

**Comment by yitam** [comment]
Yes @v-kaywon this is the error message:
PDO_SQLSRV_ERROR_INVALID_OUTPUT_PARAM_TYPE,
`An invalid PHP type was specified as an output parameter. DateTime objects, NULL values, and streams cannot be specified as output parameters.`

**Pull request merged**